File: client/src/main/java/au/com/codeka/warworlds/client/opengl/Camera.java
package au.com.codeka.warworlds.client.opengl;

import android.opengl.Matrix;
import android.view.animation.DecelerateInterpolator;

import androidx.annotation.Nullable;

/**
 * The {@link Camera} can be used to draw a {@link Scene}, scroll and zoom around.
 */
public class Camera {
  public interface CameraUpdateListener {
    /**
     * Called when the camera is translated.
     *
     * @param x The total distance the camera is translated away from the origin in the X direction.
     * @param y The total distance the camera is translated away from the origin in the Y direction.
     * @param dx The delta the camera has moved in the X direction.
     * @param dy The delta the camera has moved in the Y direction.
     */
    void onCameraTranslate(float x, float y, float dx, float dy);
  }

  private final float[] projMatrix = new float[16];
  private final float[] viewMatrix = new float[16];
  private final float[] viewProjMatrix = new float[16];
  private final float[] translateHelper = new float[16];
  private float zoomAmount;
  private boolean flinging;
  private float flingX;
  private float flingY;
  private float flingFactor;
  private DecelerateInterpolator decelerateInterpolator;
  private float screenWidth;
  private float screenHeight;
  private float translateX;
  private float translateY;
  @Nullable private CameraUpdateListener listener;

  public Camera() {
    decelerateInterpolator = new DecelerateInterpolator(1.0f);
    Matrix.setIdentityM(viewMatrix, 0);
  }

  public void setCameraUpdateListener(CameraUpdateListener listener) {
    this.listener = listener;
  }

  public void onSurfaceChanged(float width, float height) {
    Matrix.orthoM(projMatrix, 0, -width, width, -height, height, 10, -10);
    zoomAmount = 1.0f;
    flinging = false;
    screenWidth = width;
    screenHeight = height;
    zoom(2.0f);
  }

  public float getScreenWidth() {
    return screenWidth;
  }

  public float getScreenHeight() {
    return screenHeight;
  }

  public float getZoomAmount() {
    return zoomAmount;
  }

  public void onDraw() {
    if (flinging) {
      float dt = 0.016f;
      flingFactor += dt * 2.0f;
      if (flingFactor > 1.0f) {
        flinging = false;
      } else {
        float factor = 1.0f - decelerateInterpolator.getInterpolation(flingFactor);
        translate(flingX * factor * dt, flingY * factor * dt);
      }
    }
  }

  public float[] getViewProjMatrix() {
    Matrix.setIdentityM(viewMatrix, 0);
    Matrix.scaleM(viewMatrix, 0, zoomAmount, zoomAmount, 1.0f);
    Matrix.translateM(viewMatrix, 0, translateX, translateY, 0.0f);
    Matrix.multiplyMM(viewProjMatrix, 0, projMatrix, 0, viewMatrix, 0);
    return viewProjMatrix;
  }

  public void translate(float x, float y) {
    translate(x, y, false);
  }

  /**
   * Translate by the given amount.
   *
   * @param x Amount to translate in X direction.
   * @param y Amount to translate in Y direction.
   * @param silent If true, we will not call the listener.
   */
  public void translate(float x, float y, boolean silent) {
    x /= (zoomAmount * 0.5f);
    y /= (zoomAmount * 0.5f);
    translateX += x;
    translateY += y;
    if (listener != null && !silent) {
      listener.onCameraTranslate(-translateX, translateY, x, y);
    }
  }

  /**
   * Simliar to {@link #translate(float, float, boolean)} except that we translate to an absolute
   * X,Y coordinate.
   *
   * @param x The X-coordinate you want to warp to.
   * @param y The Y-coordinate you want to warp to.
   */
  public void warpTo(float x, float y) {
    translateX = x;
    translateY = y;
  }

  public void fling(float x, float y) {
    flinging = true;
    flingX = x;
    flingY = y;
    flingFactor = 0.0f;
  }

  public void zoom(float factor) {
    zoomAmount *= factor;
    if (zoomAmount > 5.0f) {
      zoomAmount = 5.0f;
    }
    if (zoomAmount < 0.25f) {
      zoomAmount = 0.25f;
    }
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/opengl/Camera.kt
package au.com.codeka.warworlds.client.opengl

import android.opengl.Matrix
import android.view.animation.DecelerateInterpolator
import au.com.codeka.warworlds.client.game.starfield.scene.StarfieldManager

/** The [Camera] can be used to draw a [Scene], scroll and zoom around. */
class Camera {
  interface CameraUpdateListener {
    /**
     * Called when the camera is translated.
     *
     * @param x The total distance the camera is translated away from the origin in the X direction.
     * @param y The total distance the camera is translated away from the origin in the Y direction.
     * @param dx The delta the camera has moved in the X direction.
     * @param dy The delta the camera has moved in the Y direction.
     */
    fun onCameraTranslate(x: Float, y: Float, dx: Float, dy: Float)
  }

  private val projMatrix = FloatArray(16)
  private val viewMatrix = FloatArray(16)

  private var viewProjMatrixDirty = true
  val viewProjMatrix = FloatArray(16)
    get() {
      if (viewProjMatrixDirty) {
        Matrix.setIdentityM(viewMatrix, 0)
        Matrix.scaleM(viewMatrix, 0, zoomAmount, zoomAmount, 1.0f)
        Matrix.translateM(viewMatrix, 0, translateX, translateY, 0.0f)
        Matrix.multiplyMM(field, 0, projMatrix, 0, viewMatrix, 0)

        Matrix.invertM(inverseViewProjMatrix, 0, field, 0)
      }

      return field
    }

  /**
   * The inverse of the viewProjMatrix, useful for projecting from screen coordinates to world
   * coordinates, for example.
   */
  var inverseViewProjMatrix = FloatArray(16)
    private set

  var zoomAmount = 0f
    private set
  private var flinging = false
  private var flingX = 0f
  private var flingY = 0f
  private var flingFactor = 0f
  private val decelerateInterpolator = DecelerateInterpolator(1.0f)
  var screenWidth = 0f
    private set
  var screenHeight = 0f
    private set
  private var translateX = 0f
  private var translateY = 0f
  private var listener: CameraUpdateListener? = null

  fun setCameraUpdateListener(listener: CameraUpdateListener?) {
    this.listener = listener
  }

  fun onSurfaceChanged(width: Float, height: Float) {
    Matrix.orthoM(projMatrix, 0, -width, width, -height, height, 10f, -10f)
    zoomAmount = 1.0f
    flinging = false
    screenWidth = width
    screenHeight = height
    zoom(2.0f)
  }

  fun onDraw() {
    if (flinging) {
      val dt = 0.016f
      flingFactor += dt * 2.0f
      if (flingFactor > 1.0f) {
        flinging = false
      } else {
        val factor = 1.0f - decelerateInterpolator.getInterpolation(flingFactor)
        translate(flingX * factor * dt, flingY * factor * dt)
      }
    }
  }

  /** "Unproject" the given screen-space x,y into world coordinates. */
  fun unproject(x: Float, y: Float): FloatArray {
    var inputVector = FloatArray(4)
    inputVector[0] = ((x / screenWidth) - 0.5f) * 2.0f
    inputVector[1]  = -((y / screenHeight) - 0.5f) * 2.0f
    inputVector[2] = 0.0f
    inputVector[3] = 1.0f
    var outputVector = FloatArray(4)
    Matrix.multiplyMV(outputVector, 0, inverseViewProjMatrix, 0, inputVector, 0)
    return outputVector
  }

  /**
   * Translate by the given amount.
   *
   * @param x Amount to translate in X direction.
   * @param y Amount to translate in Y direction.
   * @param silent If true, we will not call the listener.
   */
  fun translate(x: Float, y: Float, silent: Boolean = false) {
    val scaledX = x / (zoomAmount * 0.5f)
    val scaledY = y / (zoomAmount * 0.5f)
    translateX += scaledX
    translateY += scaledY
    viewProjMatrixDirty = true
    listener?.let {
      if (!silent) {
        it.onCameraTranslate(-translateX, translateY, scaledX, scaledY)
      }
    }
  }

  /**
   * Similar to [.translate] except that we translate to an absolute
   * X,Y coordinate.
   *
   * @param x The X-coordinate you want to warp to.
   * @param y The Y-coordinate you want to warp to.
   */
  fun warpTo(x: Float, y: Float) {
    translateX = x
    translateY = y
    viewProjMatrixDirty = true
  }

  fun fling(x: Float, y: Float) {
    flinging = true
    flingX = x
    flingY = y
    flingFactor = 0.0f
  }

  fun zoom(factor: Float) {
    zoomAmount *= factor
    if (zoomAmount > 5.0f) {
      zoomAmount = 5.0f
    }
    if (zoomAmount < 0.25f) {
      zoomAmount = 0.25f
    }
    viewProjMatrixDirty = true
  }

  init {
    Matrix.setIdentityM(viewMatrix, 0)
  }
}